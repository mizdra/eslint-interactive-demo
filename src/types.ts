export type FormData = {
  query: string;
};

export type CursorPosition = {
  styleTop: number;
  styleLeft: number;
};

/**
 * `presetIcons` オプションの要素の型。
 * 要素の型には以下が利用できる。
 * - `string`
 *   - ページのパス (例: `'page'`, `'/other-project/page'`) を表す
 * - `string[]`
 *   - ページのパスの配列 (例: `['page1', 'page2']`)
 * - `() => PresetIconsItem[]`
 *   - 動的に `PresetIconsItem[]` を生成して返す関数
 * - `() => Promise<PresetIconsItem[]>`
 *   - `() => PresetIconsItem[]` の非同期版
 * */
export type PresetIconsItem = string | string[] | (() => PresetIconsItem[]) | (() => Promise<PresetIconsItem[]>);
