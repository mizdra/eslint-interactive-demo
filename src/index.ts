export { registerIconSuggestion } from './lib/register';
export { Icon } from './lib/icon';
export { getMemberIcons } from './lib/preset-icon';
