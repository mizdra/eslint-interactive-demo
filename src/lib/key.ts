/** IME による変換中かどうか */
export function isComposing(e: KeyboardEvent): boolean {
  return e.isComposing || (e.key === 'Enter' && e.which === 229);
}
