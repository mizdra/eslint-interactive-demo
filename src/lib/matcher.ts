import Asearch from 'asearch';
import { Item } from '../types';
import { uniqBy } from './collection';

/**
 * 曖昧検索による matcher。
 * 曖昧検索で `items` をフィルタしつつ、編集距離の昇順で並べ替えて返す。
 */
export function fuzzyMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  // query の長さが 1〜3文字までは 0 文字まで、4〜7文字までは 1 文字まで、
  // 8〜11文字までは 2 文字まで、12文字以降は 3 文字まで誤字を許容する
  const maxAambig = Math.min(Math.floor(query.length / 4), 3);
  const match = Asearch(` ${query} `); // 部分一致できるように、両端をスペースで囲む
  // あいまい度の少ない項目から順に並べる (ただし重複は除く)
  const pushedItems: Set<Item<T>> = new Set();
  const result: Item<T>[] = [];
  for (let ambig = 0; ambig <= maxAambig; ambig++) {
    items.forEach((item) => {
      if (!match(item.searchableText, ambig)) return;
      if (pushedItems.has(item)) return;
      pushedItems.add(item);
      result.push(item);
    });
  }

  return result;
}

/**
 * 前方一致による matcher。
 * 元の `items` の順序を維持しつつ、`items` を前方一致でフィルタして返す。
 * */
export function startsWithMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  return items.filter((item) => item.searchableText.toLowerCase().startsWith(query.toLowerCase()));
}

/**
 * 部分一致による matcher。
 * 元の `items` の順序を維持しつつ、`items` を部分一致でフィルタして返す。
 * */
export function includesMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  return items.filter((item) => item.searchableText.toLowerCase().includes(query.toLowerCase()));
}

/**
 * 前方一致、部分一致、曖昧検索を組み合わせた matcher。
 * 前方一致 > 部分一致 > 曖昧検索 の順で fallback しながらフィルタしていき、残ったものを返す。
 * */
export function combinedMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  const newItems = [
    ...startsWithMatcher(query, items),
    ...includesMatcher(query, items),
    ...fuzzyMatcher(query, items),
  ];
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return uniqBy(newItems, (item) => item.key);
}
