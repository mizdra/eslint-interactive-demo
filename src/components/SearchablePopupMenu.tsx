import { ComponentChild } from 'preact';
import { useCallback, useEffect, useMemo, useState } from 'preact/hooks';
import { Icon } from '../lib/icon';
import { forwardPartialFuzzyMatcher } from '../lib/matcher';
import { CursorPosition, Matcher } from '../types';
import { PopupMenu } from './PopupMenu';
import { SearchInput } from './SearchablePopupMenu/SearchInput';

export type SearchablePopupMenuProps<T> = {
  open: boolean;
  emptyMessage?: string;
  icons: Icon[];
  cursorPosition: CursorPosition;
  matcher?: Matcher<T>;
  onSelect?: (icon: Icon) => void;
  onClose?: () => void;
  onInputQuery?: (query: string) => void;
  isSuggestionCloseKeyDown?: (e: KeyboardEvent) => boolean;
};

export function SearchablePopupMenu<T>({
  open,
  emptyMessage,
  icons,
  cursorPosition,
  matcher = forwardPartialFuzzyMatcher,
  onSelect,
  onClose,
  onInputQuery,
  isSuggestionCloseKeyDown,
}: SearchablePopupMenuProps<T>) {
  const [query, setQuery] = useState('');
  const matchedIcons = useMemo(() => matcher(query, icons), [icons, matcher, query]);
  const matchedIconsForPopupMenu = useMemo(
    () =>
      matchedIcons.map((icon) => ({
        key: icon.key,
        element: icon.element,
      })),
    [matchedIcons],
  );

  useEffect(() => {
    if (open === false) setQuery('');
  }, [open]);

  const handleSelect = useCallback(
    (_icon: ComponentChild, index: number) => {
      onSelect?.(matchedIcons[index]);
    },
    [matchedIcons, onSelect],
  );
  const handleClose = useCallback(() => {
    onClose?.();
  }, [onClose]);
  const handleInputQuery = useCallback(
    (query: string) => {
      setQuery(query);
      onInputQuery?.(query);
    },
    [onInputQuery],
  );

  return (
    <div>
      <PopupMenu
        open={open}
        emptyMessage={emptyMessage}
        icons={matchedIconsForPopupMenu}
        cursorPosition={cursorPosition}
        onSelect={handleSelect}
        onClose={handleClose}
        isPopupCloseKeyDown={isSuggestionCloseKeyDown}
      />
      {open && (
        <SearchInput
          defaultQuery={query}
          cursorPosition={cursorPosition}
          onInput={handleInputQuery}
          onBlur={handleClose}
        />
      )}
    </div>
  );
}
