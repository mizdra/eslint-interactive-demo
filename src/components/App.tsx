import { FunctionComponent } from 'preact';
import { useCallback, useState } from 'preact/hooks';
import { forwardPartialFuzzyMatcher } from '..';
import { useDocumentEventListener } from '../hooks/useDocumentEventListener';
import { useScrapbox } from '../hooks/useScrapbox';
import { uniqueIcons } from '../lib/collection';
import { Icon } from '../lib/icon';
import { isComposing } from '../lib/key';
import { calcCursorPosition, insertText, scanEmbeddedIcons } from '../lib/scrapbox';
import { CursorPosition, Matcher } from '../types';
import { SearchablePopupMenu } from './SearchablePopupMenu';

const DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN = (e: KeyboardEvent) => {
  return e.key === 'l' && e.ctrlKey && !e.shiftKey && !e.altKey && !e.metaKey;
};

const DEFAULT_IS_INSERT_QUERY_KEY_DOWN = (e: KeyboardEvent) => {
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && e.altKey && !e.metaKey) return true;
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && !e.altKey && e.metaKey) return true;
  return false;
};

export type AppProps = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  isSuggestionCloseKeyDown?: (e: KeyboardEvent) => boolean;
  isInsertQueryKeyDown?: (e: KeyboardEvent) => boolean;
  presetIcons?: Icon[];
  defaultSuggestPresetIcons?: boolean;
  matcher?: Matcher;
};

export const App: FunctionComponent<AppProps> = ({
  isSuggestionOpenKeyDown = DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN,
  isSuggestionCloseKeyDown,
  isInsertQueryKeyDown = DEFAULT_IS_INSERT_QUERY_KEY_DOWN,
  presetIcons = [],
  defaultSuggestPresetIcons = true,
  matcher = forwardPartialFuzzyMatcher,
}) => {
  const { textInput, cursor, editor, layout, projectName } = useScrapbox();

  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [embeddedIcons, setEmbeddedIcons] = useState<Icon[]>([]);
  const [suggestPresetIcons, setSuggestPresetIcons] = useState(defaultSuggestPresetIcons);
  const composedMatcher = useCallback(
    (query: string) => {
      const composedIcons = uniqueIcons(suggestPresetIcons ? [...embeddedIcons, ...presetIcons] : embeddedIcons);
      return matcher({ query, composedIcons, presetIcons, embeddedIcons });
    },
    [embeddedIcons, matcher, presetIcons, suggestPresetIcons],
  );
  const [query, setQuery] = useState('');

  const handleSelect = useCallback(
    (icon: Icon) => {
      setOpen(false);
      insertText(textInput, icon.getNotation(projectName));
    },
    [projectName, textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleSuggestionOpenKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      e.preventDefault();
      e.stopPropagation();

      if (!open) {
        // ポップアップが閉じていたら開く
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanEmbeddedIcons` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        // 画像化されたらエディタを走査してアイコンを収集
        const newEmbeddedIcons = scanEmbeddedIcons(projectName, editor);

        setEmbeddedIcons(newEmbeddedIcons);
        setOpen(true);
        setSuggestPresetIcons(defaultSuggestPresetIcons);
      } else {
        // ポップアップが開いていたら、preset icon の表示・非表示をトグルする
        setSuggestPresetIcons((suggestPresetIcons) => !suggestPresetIcons);
      }
    },
    [cursor, defaultSuggestPresetIcons, editor, open, layout, projectName, textInput],
  );

  const handleInsertQueryKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      if (!open) return; // ポップアップが閉じていたら無視する
      e.preventDefault();
      e.stopPropagation();
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [layout, open, textInput, query],
  );

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (isComposing(e)) return; // IMEによる変換中は何もしない
      if (isSuggestionOpenKeyDown(e)) {
        handleSuggestionOpenKeyDown(e);
      } else if (isInsertQueryKeyDown(e)) {
        handleInsertQueryKeyDown(e);
      }
    },
    [isSuggestionOpenKeyDown, isInsertQueryKeyDown, handleSuggestionOpenKeyDown, handleInsertQueryKeyDown],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SearchablePopupMenu
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      cursorPosition={cursorPosition}
      matcher={composedMatcher}
      onSelect={handleSelect}
      onClose={handleClose}
      onInputQuery={setQuery}
      isSuggestionCloseKeyDown={isSuggestionCloseKeyDown}
    />
  );
};
