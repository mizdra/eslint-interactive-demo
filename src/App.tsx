import { FunctionComponent } from 'preact';
import { useCallback, useMemo, useState } from 'preact/hooks';
import { SuggestionBox, Item } from './components/SuggestionBox';
import { useDocumentEventListener } from './hooks/useDocumentEventListener';
import { uniqBy } from './lib/collection';
import { calcCursorPosition, insertText, scanIconsFromEditor } from './lib/scrapbox';
import { CursorPosition, Icon } from './types';

const cursor = document.querySelector<HTMLElement>('.cursor')!;

function uniqueIcons(icons: Icon[]): Icon[] {
  return uniqBy(icons, (icon) => icon.pagePath);
}

function generateItem(icon: Icon): Item {
  return {
    element: (
      <span>
        <img
          alt={icon.imgAlt}
          title={icon.imgTitle}
          style="width: 1.3em; height: 1.3em; object-fit: contain;"
          src={icon.imgSrc}
        />
        {' ' + icon.pagePath}
      </span>
    ),
    searchableText: icon.pagePath,
  };
}

type AppProps = {
  isSuggestionOpenKeyDown: (e: KeyboardEvent) => boolean;
  presetIcons: Icon[];
  editor: HTMLElement;
  textInput: HTMLTextAreaElement;
};

export const App: FunctionComponent<AppProps> = ({ isSuggestionOpenKeyDown, presetIcons, editor, textInput }) => {
  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [suggestibleIcons, setSuggestibleIcons] = useState<Icon[]>([]);
  const suggestBoxItems = useMemo(() => suggestibleIcons.map(generateItem), [suggestibleIcons]);
  const [presetAppended, setPresetAppended] = useState(false);

  const handleSelect = useCallback(
    (_item: Item, index: number) => {
      setOpen(false);
      insertText(textInput, suggestibleIcons[index].notation);
    },
    [suggestibleIcons, textInput],
  );

  const handleSelectNonexistent = useCallback(
    (query: string) => {
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (!isSuggestionOpenKeyDown(e)) return;
      e.preventDefault();
      e.stopPropagation();

      if (open && !presetAppended) {
        setSuggestibleIcons(uniqueIcons([...suggestibleIcons, ...presetIcons]));
        setPresetAppended(true);
      } else {
        const icons = scanIconsFromEditor(scrapbox.Project.name, editor);
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanIconsFromEditor` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        setSuggestibleIcons(uniqueIcons(icons));
        setOpen(true);
        setPresetAppended(false);
      }
    },
    [editor, isSuggestionOpenKeyDown, open, presetAppended, presetIcons, suggestibleIcons, textInput],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SuggestionBox
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      items={suggestBoxItems}
      cursorPosition={cursorPosition}
      onSelect={handleSelect}
      onSelectNonexistent={handleSelectNonexistent}
      onClose={handleClose}
    />
  );
};
