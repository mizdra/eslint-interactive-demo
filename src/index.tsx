import { render } from 'preact';
import { App } from './App';
import { pagePathToIcon } from './lib/icon';
import { getEditor } from './lib/scrapbox';

type Options = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  // `string` is `pagePath`
  presetIcons?: string[];
  editor?: HTMLElement;
  textInput?: HTMLTextAreaElement;
  cursor?: HTMLElement;
  scrapbox?: Scrapbox;
};

export function registerIconSuggestion(options?: Options) {
  // 直接 editor に mount すると scrapbox 側の react renderer と干渉して壊れるので、
  // editor 内に差し込んだ container に mount する
  const container = document.createElement('div');
  const editor = options?.editor ?? getEditor();
  editor.appendChild(container);

  const presetIcons = options?.presetIcons?.map((pagePath) => pagePathToIcon(scrapbox.Project.name, pagePath));

  render(
    <App
      isSuggestionOpenKeyDown={options?.isSuggestionOpenKeyDown}
      presetIcons={presetIcons}
      editor={editor}
      textInput={options?.textInput}
      cursor={options?.cursor}
      scrapbox={options?.scrapbox}
    />,
    container,
  );
}
