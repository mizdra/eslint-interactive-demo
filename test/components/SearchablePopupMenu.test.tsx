import { act, fireEvent, render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { datatype } from 'faker';
import { SearchablePopupMenu } from '../../src/components/SearchablePopupMenu';
import { forwardMatcher } from '../../src/lib/matcher';
import { CursorPosition } from '../../src/types';
import { keydownEnterEvent, keydownEscapeEvent } from '../helpers/key';

// ダミーの props
const cursorPosition: CursorPosition = { styleTop: 0, styleLeft: 0 };
const items = [
  { key: 1, element: <span key="1">a</span>, searchableText: 'a', value: 'a' },
  { key: 2, element: <span key="2">ab</span>, searchableText: 'ab', value: 'ab' },
  { key: 3, element: <span key="3">abc</span>, searchableText: 'abc', value: 'abc' },
  { key: 4, element: <span key="4">z</span>, searchableText: 'z', value: 'z' },
];
const matcher = forwardMatcher;
const props = { cursorPosition, items, matcher };

describe('SearchablePopupMenu', () => {
  describe('open === false の時', () => {
    test('ポップアップも SearchInput も表示されない', () => {
      const { asFragment, queryByTestId } = render(<SearchablePopupMenu open={false} {...props} />);
      expect(queryByTestId('popup-menu')).toBeNull();
      expect(queryByTestId('search-input')).toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('open === true の時', () => {
    test('ポップアップと SearchInput が表示される', () => {
      const { asFragment, queryByTestId } = render(<SearchablePopupMenu open {...props} />);
      expect(queryByTestId('popup-menu')).not.toBeNull();
      expect(queryByTestId('search-input')).not.toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });
    test('Esc 押下で onClose が呼び出される', async () => {
      const onClose = jest.fn();
      render(<SearchablePopupMenu open {...props} onClose={onClose} />);
      expect(onClose).toBeCalledTimes(0);
      await act(() => {
        fireEvent(document, keydownEscapeEvent);
      });
      expect(onClose).toBeCalledTimes(1);
    });
    describe('ポップアップに表示されるアイテムが空の時', () => {
      test('emptyMessage でアイテムが空の時のメッセージを変更できる', () => {
        const emptyMessage = datatype.string();
        const { getByText } = render(<SearchablePopupMenu open {...props} items={[]} emptyMessage={emptyMessage} />);
        expect(getByText(emptyMessage)).toBeInTheDocument();
      });
    });
    describe('ポップアップに表示されるアイテムが空でない時', () => {
      test('SearchInput に文字を入力するとアイテムがフィルタされる', () => {
        const { getByTestId } = render(<SearchablePopupMenu open {...props} />);
        const buttonContainer = getByTestId('button-container');
        const searchInput = getByTestId('search-input');

        expect(buttonContainer.childElementCount).toEqual(4);
        userEvent.type(searchInput, 'a');
        expect(buttonContainer.childElementCount).toEqual(3);
        userEvent.type(searchInput, 'b');
        expect(buttonContainer.childElementCount).toEqual(2);
        userEvent.type(searchInput, 'c');
        expect(buttonContainer.childElementCount).toEqual(1);
        userEvent.type(searchInput, 'd');
        expect(buttonContainer.childElementCount).toEqual(0);
      });
      test('Enter 押下で onSelect が呼び出される', async () => {
        const onSelect = jest.fn();
        render(<SearchablePopupMenu open {...props} onSelect={onSelect} />);
        expect(onSelect).toBeCalledTimes(0);
        await act(() => {
          fireEvent(document, keydownEnterEvent);
        });
        expect(onSelect).toBeCalledTimes(1);
      });
    });
  });
  test('open === true になった時に、 SearchInput に入力された文字がリセットされる', () => {
    const { rerender, getByTestId } = render(<SearchablePopupMenu open {...props} />);

    userEvent.type(getByTestId('search-input'), 'a');
    expect(getByTestId('search-input')).toHaveValue('a');

    rerender(<SearchablePopupMenu open={false} {...props} />);
    rerender(<SearchablePopupMenu open {...props} />);

    expect(getByTestId('search-input')).toHaveValue('');
  });
});
