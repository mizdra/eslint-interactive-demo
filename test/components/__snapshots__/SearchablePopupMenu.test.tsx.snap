// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchablePopupMenu open === false の時 ポップアップも SearchInput も表示されない 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`SearchablePopupMenu open === true の時 ポップアップと SearchInput が表示される 1`] = `
<DocumentFragment>
  <div>
    <div
      class="popup-menu"
      data-testid="popup-menu"
      style="top: 0px;"
    >
      <div
        class="button-container"
        data-testid="button-container"
        style="left: 0px; transform: translateX(-NaN%);"
      >
        <div
          class="button selected"
        >
          <span>
            a
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            ab
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            abc
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            z
          </span>
        </div>
      </div>
      <div
        class="triangle"
        style="left: 0px;"
      />
    </div>
    <form
      class="form-inline"
    >
      <input
        class="form-control"
        data-testid="search-input"
        default="true"
        style="position: absolute; top: 0px; left: 0px; transform: translateX(-NaN%);"
      />
    </form>
  </div>
</DocumentFragment>
`;
